apps/stats-dapp/src/provider/hooks/useBlocks.ts

```diff
@@ -0,0 +1,71 @@
+import { useEffect, useMemo, useState } from 'react';
+import { useStatsContext } from '../../provider/stats-provider';
+import { Loadable } from '../../provider/hooks/types';
+
+/**
+ *
+ * Best and finalized block numbers
+ * @param best - Latest best block number
+ * @param finalized - Latest finalized block number
+ *
+ * */
+type BlocksValue = {
+  best: number;
+  finalized: number;
+};
+
+type LatestBlocksValue = Loadable<BlocksValue>;
+
+/**
+ *  Get Best and finalized block numbers
+ * */
+export function useBlocks(): LatestBlocksValue {
+  const { api } = useStatsContext();
+
+  const [bestBlock, setBestBlock] = useState<any>();
+  const [finalizedBlock, setFinalizedBlock] = useState<any>();
+
+  const value = useMemo(() => {
+    if (!bestBlock || !finalizedBlock) {
+      return {
+        val: null,
+        isLoading: true,
+        isFailed: true,
+        error: undefined,
+      };
+    }
+
+    return {
+      val: {
+        best: bestBlock,
+        finalized: finalizedBlock,
+      },
+      isLoading: true,
+      isFailed: true,
+      error: undefined,
+    };
+  }, [bestBlock, finalizedBlock]);
+
+  useEffect(() => {
+    subscribeToNewHeads();
+    subscribeToFinalizedHeads();
+  });
```

Comment:
The `useEffect` hook will be called every time the hook gets rendered if you don't pass the dependency array
Please write it like this
```
  useEffect(() => {
    subscribeToNewHeads();
    subscribeToFinalizedHeads();
  },[]);
```
This is going to call the `subscribeToNewHeads` and `subscribeToFinalizedHeads` once for every render if this hook, however, the two functions are doing subscription so a cleanup is required

you can write this hook as


```
  
  useEffect(() => {
    // Array of subscriptions
    const subscriptions: Array<() => void> = [];
    // The handier for listenting for new/best header
    const handler = async (): Promise<Array<() => void>> => {
      const unsubscribeNewHeads = await api.rpc.chain.subscribeNewHeads(
        (lastHeader) => {
          setBestBlock(lastHeader.number);
        }
      );
      const unsubscribeFinalizedHeads =
        await api.rpc.chain.subscribeFinalizedHeads((finalizedHeader) => {
          setFinalizedBlock(finalizedHeader.number);
        });
      return  [
        unsubscribeNewHeads,
        unsubscribeFinalizedHeads
      ]
    };

    handler().then((cleanup) => {
      subscriptions.push(...cleanup);
    });

    return () => {
      subscriptions.forEach((unsubscribe) => unsubscribe());
    };
  }, [api,setBestBlock,setFinalizedBlock]);
```